broadinstitute/scp-ingest-pipeline
# Log ingest annotation validation errors to Mixpanel (SCP-3921)

Adding Mixpanel logging to ingest pipeline errors. This PR addresses logging of errors that use the store_validation_issue from the Annotation Class. Logging expression matrix ingest errors requires a different approach and will be addressed in follow on ticket SCP-4037.

This PR adds a "file-validation" event to Mixpanel logging upon completion of an ingest pipeline run. The event has the following properties ingest pipeline has not reported before:

- logger - always set to "ingest-pipeline"
- trigger - "sync" if study file has remote_location set
                "dev_mode" if BYPASS_MONGO_WRITES is set to "yes"
                "upload" if neither of the above are true
- status - success|failure of the ingest pipeline job
- errorTypes - aligned with errorType used in CSFV
- errors - array of error messages sent to user
- numErrorTypes
- numErrors

This PR adds the option to bypass input validations and writes to MongoDB for development work. In the scripts directory, example script `setup_mongo_dev` sets up environment variables that:
   - set up access to remote Mongo instance
   - BYPASS_MONGO_WRITES='yes' bypasses input validation and writes
   - BARD_HOST_URL set to log events to Mixpanel dev
   If mixpanel logging during development is not needed, please:
      unset BARD_HOST_URL
	
To test:

1. via command line, set up to run in developer mode (can use `setup_mongo_dev` from scripts directory).
2. from ingest directory, run this command the indicated scp-ingest-pipeline repo file or a file of your choice:
python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 ingest_cell_metadata --cell-metadata-file ../tests/data/annotation/metadata/convention/has_na_in_required_fields.csv --study-accession SCPdev --ingest-cell-metadata --validate-convention

Check for the following CLI results and confirm they are reflected in the file-validation event in Mixpanel Dev:
  logger: ingest-pipeline
  trigger: dev_mode
  appId: single-cell-portal
  errorTypes: ['content:metadata:missing-required-values', 'content:schema-error']
  errors: ['organ: required column "organ" missing data', 'species: required column "species__ontology_label" missing data', 'One or more errors from jsonschema validation of metadata content against Alexandria Convention detected']
  numErrortypes: 2
  numErrors: 3
  functionName: ingest_cell_metadata
  status: failure

Note: the following are dummy values hard-coded when using dev_mode (no info from Mongo on studyFile):
  distinct_id: 2f30ec50-a04d-4d43-8fd1-b136a2045079
  studyAccession: SCPdev
  fileType: input_validation_bypassed
  fileSize: 1
  trigger: dev_mode

(optional) Additional validation of trigger functionality

Test for trigger: upload
1. set your local instance to use docker image:
`gcr.io/broad-singlecellportal-staging/scp-ingest-jlc_mixpanel_logging:623864e`
2. upload a file that will pass CSFV and confirm that the mixpanel file-validation event with logger: ingest-pipeline shows trigger: upload

Test for trigger: sync
1. using above docker image, gsutil a file to a study bucket for a study in your local dev environment
2. sync the file and confirm that the mixpanel file-validation event with logger: ingest-pipeline shows trigger: sync

This PR support SCP-3921
